### ISSUE TITLE
huge FPS penalty running autounsuspend

### ISSUE THREAD

**Issue by rrauenza** [DFHack/dfhack#1076]
I was down to 1 FPS and have been playing with my fortress trying to figure out what might be causing it.  I think in part it's caused by a forbidden beast I walled off.

But I also just shut off autounsuspend (# autosuspend stop) and I found my FPS went from 1 to 15.  I have lots of suspended jobs from the workflow hack (e.g., create 5-10 beds) ...so I wonder if there was a conflict.  I can reliably reproduce this.

What's also kind of odd is that I've had autounsuspend on for a while and it has only recently been an issue in the last couple of days.

I'm fine with just leaving it off, but I wanted to give a heads up because I know FPS is a difficult issue to troubleshoot.
```

[DFHack]# autounsuspend resume
Stopped.
Resuming job 614840: MakeWeapon [steel] (delayed)
Resuming job 614843: MakeWeapon [steel] (delayed)
Resuming job 614844: MakeWeapon [steel] (delayed)
Resuming job 855769: MakePotashFromLye [potash] (delayed)
Resuming job 893625: MakePotashFromLye [potash] (delayed)
Resuming job 893626: MakePotashFromAsh [potash] (delayed)
Resuming job 1038734: MakeAmmo [iron] (delayed)
Resuming job 1064322: MakePants [copper] (delayed)
Resuming job 1064324: MakePants [steel] (delayed)
[DFHack]# autounsuspend resume
Stopped.
Resuming job 1535: BREW_DRINK_FROM_PLANT
Resuming job 397153: BREW_DRINK_FROM_PLANT
Resuming job 622635: BREW_DRINK_FROM_PLANT
Resuming job 855770: MakePotashFromAsh [potash] (delayed)
Resuming job 871829: MakeArmor [cloth] (delayed)
Resuming job 893634: MakePotashFromLye [potash] (delayed)
Resuming job 893635: MakePotashFromAsh [potash] (delayed)
Resuming job 905253: BREW_DRINK_FROM_PLANT
Resuming job 905254: BREW_DRINK_FROM_PLANT_GROWTH
Resuming job 1064323: MakePants [bronze] (delayed)
Resuming job 1064325: MakePants [bismuth bronze] (delayed)
Resuming job 1082616: MeltMetalObject [metal] (delayed)
Resuming job 1193925: MakeTrapComponent [copper] (delayed)
[DFHack]# autounsuspend start
[DFHack]# autounsuspend stop
[DFHack]# autounsuspend resume
Stopped.
[DFHack]#

```

**Comment by lethosor** [comment]
autounsuspend checks all of the jobs periodically. It looks like your highest job ID is 1193925, which is *really* high - either an indicator that you have a ton of jobs (maybe due to large designations) or something else has gone wrong.

This should tell you how many jobs are present - mind running it in the console?
```
:lua i=0 j=world.job_list while j.next do i=i+1 j=j.next end print(i)
```

**Comment by rrauenza** [comment]
...So the code unsuspends every 5 ticks if I read it right.   That seems kind of frequent -- shouldn't it really need to run no more than once a second (real time).   At 50 I think (if I understand ticks right), it would trigger once a dwarf hour, which at 100FPS is twice a real second?

**Comment by rrauenza** [comment]
....nothing.  

```
[DFHack]# lua i=0 j=world.job_list while j.next do i=i+1 j=j.next end print(i)
[DFHack]#
```

**Comment by lethosor** [comment]
You forgot the colon in the command I gave you.

One day is 1200 ticks, so that's about right. You could increase the number in `df.onupdate_register('autounsuspend', 5)` and see if that helps.

**Comment by rrauenza** [comment]
Oh yeah I dropped the colon because it gave an error :)
```

[DFHack]# :lua i=0 j=world.job_list while j.next do i=i+1 j=j.next end print(i)
(lua command):1: attempt to index global 'world' (a nil value)
stack traceback:
        (lua command):1: in main chunk
        [C]: in function 'safecall'
        ...PERIDE~1.03-\Dwarf Fortress 0.43.03/hack/scripts/lua.lua:79: in function 'f'
        ...\PERIDE~1.03-\Dwarf Fortress 0.43.03\hack\lua\dfhack.lua:562: in function <...\PERIDE~1.03-\Dwarf Fortress 0.43.03\hack\lua\dfhack.lua:503>
        (...tail calls...)
```

**Comment by lethosor** [comment]
What DFHack version are you using? I think the issue is that you're using something before 0.43.05-beta1, not the colon.

Try this instead:
```
:lua i=0 j=df.global.world.job_list while j.next do i=i+1 j=j.next end print(i)
```

**Comment by rrauenza** [comment]
PeridexisErrant's Starter Pack 0.43.03-r09

[lua]# i=0 j=df.global.world.job_list while j.next do i=i+1 j=j.next end print(i)
252543

(Now how do I drop back out of the lua interpreter?)

**Comment by lethosor** [comment]
I was looking for the DFHack version, although I'm guessing that's "0.43.03-r1". "quit" should get you out of the Lua interpreter.

Anyway, 252543 jobs is a lot. Are there that many listed on the jobs screen? If not, do you have a lot of designations?

**Comment by rrauenza** [comment]
I've got two z-levels designated with the hope that the forgotten beast will be found.  The problem existed before designating those zlevels, but let me undesignate them.

**Comment by rrauenza** [comment]
How can I see which dfhack is bundled?

**Comment by lethosor** [comment]
I suppose designating two entire z-levels on a 7x7 embark could do it. I'm not sure if there's a faster way to check just the jobs visible in the job list (which are presumably the ones autounsuspend should focus on) or not.

The DFHack version should be on the title screen, and logged in the DFHack console when DF starts. It should ideally also be in your pack's documentation somewhere. Otherwise, `:lua print(dfhack.getDFHackVersion())` will tell you.

**Comment by rrauenza** [comment]
0.43.03-r1

Huh.. I did digexp clear on both of those z's ... but...

# :lua i=0 j=df.global.world.job_list while j.next do i=i+1 j=j.next end print(i)
232011

**Comment by rrauenza** [comment]
I wonder if the root of my FPS is the jobs, and it's just that autounsuspend made it more obvious.  I need to figure out a way to dump those jobs to a text file.

**Comment by lethosor** [comment]
I think the number of jobs could be a major factor, although it's not necessarily the only thing. Are there a lot listed on the [j]obs screen?

**Comment by rrauenza** [comment]
Not hundreds of thousands of jobs in the job screen.  5 screen fulls maybe.

...I'm wondering if my loom dwarf has thousands of webs to collect from the lost FB.  The reveal command showed a lot of webs, but not hundreds of thousands.

**Comment by rrauenza** [comment]
]# rb df.world.job_list.each { |job|  puts job.job_type }

Whole lot of FellTree ... turning off autochop.

**Comment by rrauenza** [comment]
# rb count = 0; df.world.job_list.each { |job|  if job.job_type == :FellTree then count += 1 end } ; puts count
228928

**Comment by rrauenza** [comment]
Trying to find a way to programmatically cancel those FellTree's ...

**Comment by rrauenza** [comment]
Looped over the jobs and printed the job.pos:

```
#<Coord x=232 y=91 z=151>
#<Coord x=233 y=82 z=151>
#<Coord x=235 y=92 z=151>
#<Coord x=235 y=95 z=151>
#<Coord x=236 y=88 z=151>
#<Coord x=237 y=81 z=151>
#<Coord x=237 y=92 z=151>
#<Coord x=237 y=95 z=151>
#<Coord x=239 y=80 z=151>
#<Coord x=239 y=85 z=151>
#<Coord x=239 y=95 z=151>
#<Coord x=224 y=106 z=151>
#<Coord x=225 y=101 z=151>
#<Coord x=226 y=97 z=151>
#<Coord x=230 y=96 z=151>
#<Coord x=230 y=99 z=151>
#<Coord x=230 y=103 z=151>
#<Coord x=230 y=109 z=151>
#<Coord x=235 y=104 z=151>
#<Coord x=235 y=108 z=151>
#<Coord x=236 y=97 z=151>
#<Coord x=238 y=99 z=151>
#<Coord x=238 y=104 z=151>
#<Coord x=239 y=105 z=151>
#<Coord x=239 y=110 z=151>
#<Coord x=237 y=114 z=151>
#<Coord x=238 y=122 z=151>
#<Coord x=239 y=114 z=151>
0
[DFHack]#
```

My Z in the game only goes up to 139 ...

**Comment by rrauenza** [comment]
Filtering out z==151, I still have a lot of other Z's ...and some repeats. These are just FellTree:

```
#<Coord x=168 y=108 z=113>
#<Coord x=161 y=112 z=113>
#<Coord x=164 y=117 z=113>
#<Coord x=165 y=119 z=113>
#<Coord x=3 y=9 z=152>
#<Coord x=7 y=7 z=152>
#<Coord x=3 y=9 z=152>
#<Coord x=7 y=7 z=152>
#<Coord x=3 y=9 z=152>
#<Coord x=7 y=7 z=152>
#<Coord x=3 y=9 z=152>
#<Coord x=7 y=7 z=152>
```

**Comment by rrauenza** [comment]
Finally figured out how to clean up the jobs.

I first walked each FellTree jobs to see if any tiles were marked for FellTree.  None were.  So these jobs existed without a corresponding tile.

I cobbled together this, some of the code taken from advFort.lua:

```
function smart_job_delete( job )
    local gref_types=df.general_ref_type
    --TODO: unmark items as in job
    for i,v in ipairs(job.general_refs) do
        if v:getType()==gref_types.BUILDING_HOLDER then
            local b=v:getBuilding()
            if b then
                --remove from building
                for i,v in ipairs(b.jobs) do
                    if v==job then
                        b.jobs:erase(i)
                        break
                    end
                end
            else
                print("Warning: building holder ref was invalid while deleting job")
            end
        elseif v:getType()==gref_types.UNIT_WORKER then
            local u=v:getUnit()
            if u then
                u.job.current_job =nil
            else
                print("Warning: unit worker ref was invalid while deleting job")
            end
        else
            print("Warning: failed to remove link from job with type:",gref_types[v:getType()])
        end
    end
    --unlink job
    local link=job.list_link
    if link.prev then
        link.prev.next=link.next
    end
    if link.next then
        link.next.prev=link.prev
    end
	-- Don't delete the ptrs because I think something still is hanging on to them.
    --link:delete()
    --finally delete the job
    --job:delete()
end

local job_link=df.global.world.job_list.next

while job_link do
    local job = job_link.item
	job_link = job_link.next
	if job and job.job_type == 9 then
		dfhack.job.printJobDetails(job)
		smart_job_delete(job)
	end
end
```

I had to comment out the parts that actually freed the memory or the save crash -- my guess is there was a double free or something.  As long as the double linked list was trimmed, the save process saved the trimmed list.

So this might be a bug with both autounsuspend (make the tick much higher by like 500-1000), and autochop seems to be adding FellTree jobs to the job queue without tile designations.

**Comment by rrauenza** [comment]
I can confirm when enabling autochop my chop jobs began spiraling again.  I had a burrow designated and enabled.   I asked for 200-400 trees.

**Comment by lethosor** [comment]
Yeah, it seems that autochop has some issues. Unfortunately, I'm not too familiar with it, so I don't know if I can fix them easily. A couple came up in #1030, which may or may not be related.

If you can run this save under multiple DF versions, it would be great to know when this issue started appearing (i.e. maybe checking 0.40.24, 0.42.06, and 0.43.05). I'm guessing you've saved this one under 0.43.03, though, which means your current save can't be backported to earlier versions. It still might be helpful to know if it happens in 0.43.05, though.

If nobody steps up who knows more about autochop within a week, I'll try my best to sort out what's going on with it (I'm fairly busy for the rest of this week).

**Comment by ab9rf** [comment]
This really has nothing to do with autounsuspend; it's entirely because autochop is spamming the living daylights out of the job queue, by constantly spawing FellTree jobs over and over again. See, e.g, #1014, #1018, #1030.

I know what the problem and how to fix it on a high level, but it's more of a fix than I'm up to attempting right now. Basically, anything that, prior to 0.40, relied on map designations to check for a dig/chop/gather job now has to _also_ check job postings (and maybe also the job list, not sure on that), and anything that wants to cancel a dig/chop/gather job has to be prepared to remove entries from the job posting list. You can still queue new jobs by setting map designations, but you should mark the map dirty by setting `df.global.process_dig` when you do. The "best" solutions would be some convenience routines in one of dfhack's modules (`Map`, I suppose) for interrogating for, starting, and cancelling such jobs, since this is something that many different plugins do.

I would strongly urge people not to use autochop until we get this issue fixed properly.

**Comment by lethosor** [comment]
I know there's a bit of code to deal with postings in the Job module (from #741), so I'd put it there. Thanks for the information! I'll look into fixing it when I get a chance, hopefully this weekend.

**Comment by lethosor** [comment]
@rrauenza Looking over this again, a couple things I noticed:
- It's possible that the Z levels reported by the game are different from the ones used internally - can you run `lua printall(df.global.cursor)` around trees and see if you get a `z` close to 151?
- I think `dfhack.job.removeJob()` accomplishes something similar to your `smart_job_delete()`

**Comment by lethosor** [comment]
The root cause of this that @ab9rf mentioned has been fixed in 1e46945 and a527091. I'll increase the delay in autosuspend as well, but this should be fixed in the next release.
